conan/tools/meson/toolchain.py

```diff
@@ -3,187 +3,169 @@
+        self.pkg_config_path = self._conanfile.generators_folder
 
         check_using_build_profile(self._conanfile)
 
-    def _get_backend(self, recipe_backend):
-        # Returns the name of the backend used by Meson
-        conanfile = self._conanfile
-        # Downstream consumer always higher priority
-        backend_conf = conanfile.conf["tools.meson.mesontoolchain:backend"]
-        if backend_conf:
-            return backend_conf
-        # second priority: the recipe one:
-        if recipe_backend:
-            return recipe_backend
-        # if not defined, deduce automatically the default one (ninja)
-        return 'ninja'
-
-    @staticmethod
-    def _to_meson_value(value):
-        # https://mesonbuild.com/Machine-files.html#data-types
-        import six
-
-        try:
-            from collections.abc import Iterable
-        except ImportError:
-            from collections import Iterable
-
-        if isinstance(value, six.string_types):
-            return "'%s'" % value
-        elif isinstance(value, bool):
-            return 'true' if value else "false"
-        elif isinstance(value, six.integer_types):
-            return value
-        elif isinstance(value, Iterable):
-            return '[%s]' % ', '.join([str(MesonToolchain._to_meson_value(v)) for v in value])
-        return value
-
-    @staticmethod
-    def _to_meson_build_type(build_type):
-        return {"Debug": "'debug'",
-                "Release": "'release'",
-                "MinSizeRel": "'minsize'",
-                "RelWithDebInfo": "'debugoptimized'"}.get(build_type, "'%s'" % build_type)
-    # FIXME : use 'custom' otherwise? or use just None?
-
-    @property
-    def _debug(self):
-        return self._build_type == "Debug"
-
-    @property
-    def _ndebug(self):
-        # ERROR: Value "True" (of type "boolean") for combo option "Disable asserts" is not one of
-        # the choices. Possible choices are (as string): "true", "false", "if-release".
-        return "true" if self._build_type != "Debug" else "false"
-
-    @staticmethod
-    def _to_meson_vscrt(vscrt):
-        return {"MD": "'md'",
-                "MDd": "'mdd'",
-                "MT": "'mt'",
-                "MTd": "'mtd'"}.get(vscrt, "'none'")
+        self.cross_build = {}
+        if cross_building(conanfile):
+            os_build, arch_build, os_host, arch_host = get_cross_building_settings(self._conanfile)
+            self.cross_build["build"] = self._to_meson_machine(os_build, arch_build)
+            self.cross_build["host"] = self._to_meson_machine(os_host, arch_host)
+            if hasattr(conanfile, 'settings_target') and conanfile.settings_target:
+                settings_target = conanfile.settings_target
+                os_target = settings_target.get_safe("os")
+                arch_target = settings_target.get_safe("arch")
+                self.cross_build["target"] = self._to_meson_machine(os_target, arch_target)
+
+        default_comp = ""
```

Comment:
I'd like to avoid anything related to the default comp here, as much as possible.
I think we already agreed to use explicit things rather than autodetection.
many people complain that conan picks the wrong compiler executable (e.g. `/usr/bin/gcc` instead of `gcc-10` or whatever).

for CL, we can make an exception, as we know it's always `cl.exe`.
for GCC/Clang, we basically, don't know, it could be just `gcc.exe`, `gcc-10.exe`, something like `arm-linux-gnueabi-gcc.exe`...
I'd prefer to hard fail instead of silently producing incorrect binaries with the wrong compiler.

and for this kind of thing, I suppose we need to follow blocks approach, as we discuss in #9710

I beg you to exclude this kind of tentative/risky changes from this PR, and continue the discussion we started before

Reply:
This is not true. This is not auto-detection, this is a very sane default.
The information is already defined in the ``settings.compiler``, we are not auto-detecting the compiler. 

What doesn't make sense is to always force users to define a ``[buildenv] CC=gcc CXX=g++`` when ``settings.compiler=gcc``. We don't do it in any other build system integration, and it shouldn't be done in Meson either. Assuming a ``gcc/g++``, ``clang/clang++`` is a good default, and the user always has the chance to override it defining the env-var. Otherwise, it is impossible to build a Meson project in Linux or OSX with the default detected profile, and requires all users to manually add CC/CXX env-vars to it.

So this is staying, it must be possible for every build system integration to be able to build a basic project with the default detected profile, for a default system installation of compilers and build systems.

Reply:
the user might not have a `/usr/bin/gcc` or `/usr/bin/clang` at all.
for the `default` profile, it will pick `/usr/bin/gcc` for `settings.compiler=gcc`.
but it will also pick  `/usr/bin/gcc` for non default (e.g. `armv7`) profile (e.g. `settings.arch=armv7`).
that might not be something desired in this particular case, and in many others (e.g. Android NDK).
there is no check this choice is done for the `default` profile only, it always applies.
yes, it makes some use-cases more smooth, but other use-cases might be silently broken because of that default.
I'd limit these defaults to some known cases only, e.g. :
- apple-clang with XCode, where the compiler should be located via XCRun
- MSVC/Visual Studio, where VCVars is responsible for the compiler discovery
- Android NDK, where we have a clear mapping of settings to the compiler executables

for others, we don't know and can't be sure what is the sane choice. it might be `gcc.exe`, `gcc-10.exe`, `gcc-10.0.exe` or `arm-linux-gnueabi-gcc.exe`.

Reply:
> but it will also pick /usr/bin/gcc for non default (e.g. armv7) profile (e.g. settings.arch=armv7).
that might not be something desired in this particular case, and in many others (e.g. Android NDK).

Of course the default will not work in cases of cross-compiling, and it will require specifying the specific compiler you want for cross-compile. This doesn't mean that for 75% of the users, doing native builds, we shouldn't define a default. Again, this is the same default as our CMake integration, by default it will pick the system "gcc/g++", and only for cross-compile scenarios you need to specify the CC/CXX env-vars.

Reply:
how about limiting the scope for native file only?

Reply:
Are you aware that if we don't add this, it will be also challenging to build packages using Meson in ConanCenter, where we are using the default profiles? If we change all default profiles in ConanCenter, which are different to the user default profiles, the risk of things not working also increases.

Reply:
right now, in CCI, the only problematic use case is M1. all other default profiles work.

Reply:
Because the recipes there are not using MesonToolchain yet. If MesonToolchain requires defining CC/CXX in all profiles, for all builds, things will be broken, it will be not possible to update without changing all the profiles in ConanCenter. Recall that it is necessary to move all recipes to the  new integrations to be ready for 2.0.

Reply:
we already tried MesonToolchain in CCI in the past: https://github.com/conan-io/conan-center-index/pull/7360/files https://github.com/conan-io/conan-center-index/pull/6678/files.
the thing is that it requires explicit CC/CXX for cross-building.
but it doesn't require it for native builds. e.g. it is able to deduce `cl` for Visual studio and `gcc` for GCC and `clang` for XCode builds.
profiles don't need to change.
the only one we currently have trouble is M1, but for this one, it needs a bunch of other flags (like `-arch` and `-isysroot`) from the XCrun.